packages/plugin-data/src/batcher.ts

```diff
@@ -0,0 +1,303 @@
+            newBatch.on('error', (err: Error) => {
+              // reword no external id error message to direct it to org user rather than api user
+              if (err.message.startsWith('External ID was blank')) {
+                err.message = messages.getMessage('ExternalIdRequired', [sobjectType]);
+              }
+              if (err.message.startsWith('Polling time out')) {
+                const jobIdIndex = err.message.indexOf('750');
+                const batchIdIndex = err.message.indexOf('751');
+                this.ux.log(
+                  messages.getMessage('TimeOut', [
+                    err.message.substr(jobIdIndex, 18),
+                    err.message.substr(batchIdIndex, 18),
+                  ])
+                );
+              }
+              reject(err.message);
+            });
+
+            newBatch.on('queue', (): void => {
+              batchesQueued++;
+              if (batchesQueued === batches.length) {
+                const id = job.id;
+                job.close();
+                // jsforce clears out the id after close, but you should be able to close a job
+                // after the queue, so add it back so future batch.check don't fail.*/
+                job.id = id;
+              }
+            });
+
+            if (!wait) {
+              newBatch.on(
+                'queue',
+                // we're using an async method on an event listener which doesn't fit the .on method parameter types
+                // eslint-disable-next-line @typescript-eslint/no-misused-promises
+                async (batchInfo: BatchInfo): Promise<void> => {
+                  this.ux.log(messages.getMessage('CheckStatusCommand', [i + 1, batchInfo.jobId, batchInfo.id]));
+                  const result: BatchInfo = await newBatch.check();
+                  if (result.state === 'Failed') {
+                    reject(result.stateMessage);
+                  } else {
+                    resolve(batchInfo);
+                  }
+                }
+              );
+            } else {
+              resolve(this.waitForCompletion(newBatch, batchesCompleted, overallInfo, i + 1, batches.length, wait));
+            }
+
+            newBatch.execute(batch, (err) => {
+              if (err) {
+                reject(err);
+              }
+            });
+          });
+        }
+      )
+    )) as BulkResult[];
+  }
+
+  /**
+   * register completion event listeners on the batch
+   * exposed for unit testing
+   *
+   * @param newBatch
+   * @param batchesCompleted
+   * @param overallInfo
+   * @param batchNum
+   * @param totalNumBatches
+   * @param waitMins
+   */
+  private async waitForCompletion(
+    newBatch: Batch,
+    batchesCompleted: number,
+    overallInfo: boolean,
+    batchNum: number,
+    totalNumBatches: number,
+    waitMins: number
+  ): Promise<void> {
+    const logger: Logger = await Logger.child(this.constructor.name);
```

Comment:
I think we should create the logger synchronously in the constructor, or a new one will be created every time the function is called.

Reply:
Agree with Steve